File: app/src/main/java/xyz/rodit/snapmod/StoryHelper.java
package xyz.rodit.snapmod;

import java.util.Map;

import de.robv.android.xposed.XposedBridge;
import xyz.rodit.snapmod.mappings.ContextSession;
import xyz.rodit.snapmod.mappings.ContextStoryKeys;
import xyz.rodit.snapmod.mappings.FriendStoryKeys;
import xyz.rodit.snapmod.mappings.OperaMediaInfo;
import xyz.rodit.snapmod.mappings.ParamsMap;
import xyz.rodit.snapmod.mappings.PlayableStorySnap;
import xyz.rodit.snapmod.mappings.SnapUsername;
import xyz.rodit.snapmod.mappings.StoryMetadata;

public class StoryHelper {

    public static StoryMedia getMediaInfo(ParamsMap metadata) {
        Map<?, ?> map = metadata.getMap();

        String username = "unknown";
        ContextSession session = ContextSession.wrap(map.get(ContextStoryKeys.getContextSession().instance));
        SnapUsername snapUsername = session.getInfo().getUsername();
        if (snapUsername.isNotNull()) {
            username = snapUsername.getDisplayString();
        } else {
            PlayableStorySnap storySnap = PlayableStorySnap.wrap(map.get(FriendStoryKeys.getPlayableSnapStoryRecord().instance));
            if (storySnap.isNotNull()) {
                username = storySnap.getDisplayName();
            }
        }

        Object imageMedia = map.get(StoryMetadata.getImageMediaInfo().instance);
        if (imageMedia != null) {
            return new StoryMedia(OperaMediaInfo.wrap(imageMedia), username, ".jpg");
        }

        Object videoMedia = map.get(StoryMetadata.getVideoMediaInfo().instance);
        if (videoMedia != null) {
            return new StoryMedia(OperaMediaInfo.wrap(videoMedia), username, ".mp4");
        }

        Object imageOverlayMedia = map.get(StoryMetadata.getOverlayImageMediaInfo().instance);
        if (imageOverlayMedia != null) {
            return new StoryMedia(OperaMediaInfo.wrap(imageOverlayMedia), username, ".jpg");
        }

        XposedBridge.log("Error getting media info for " + metadata + ".");
        return null;
    }

    public static class StoryMedia {

        public final OperaMediaInfo info;
        public final String username;
        public final String extension;

        public StoryMedia(OperaMediaInfo info, String username, String extension) {
            this.info = info;
            this.username = username;
            this.extension = extension;
        }
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/saving/StoryHelper.kt
package xyz.rodit.snapmod.features.saving

import xyz.rodit.snapmod.createDummyProxy
import xyz.rodit.snapmod.features.FeatureContext
import xyz.rodit.snapmod.logging.XLog
import xyz.rodit.snapmod.mappings.*
import xyz.rodit.snapmod.util.PathManager
import xyz.rodit.snapmod.util.download
import xyz.rodit.snapmod.util.toSnapUUID
import xyz.rodit.snapmod.util.toUUIDString
import xyz.rodit.xposed.client.http.StreamProvider
import xyz.rodit.xposed.client.http.streams.FileProxyStreamProvider
import java.net.URL

private val mLog = XLog("StoryHelper")

typealias UsernameFetcher = (Map<*, *>) -> String?

private val usernameFetchers = listOf<UsernameFetcher>(
    {
        val session = ContextSession.wrap(it[ContextStoryKeys.getContextSession().instance])
        if (session.isNull) null else {
            val snapUsername = session.info.username
            if (snapUsername.isNotNull) snapUsername.displayString else null
        }
    },
    {
        val storySnap =
            PlayableStorySnap.wrap(it[FriendStoryKeys.getPlayableSnapStoryRecord().instance])
        if (storySnap.isNotNull) storySnap.displayName else null
    }
)

fun downloadOperaMedia(context: FeatureContext, type: String?, media: StoryMedia?) {
    if (media == null || media.info.isNull) return

    val provider: StreamProvider = FileProxyStreamProvider(context.appContext) {
        try {
            var stream = URL(media.info.uri).openStream()
            val enc = media.info.encryption
            if (enc.isNotNull) {
                stream = enc.decryptStream(stream)
            }

            return@FileProxyStreamProvider stream
        } catch (e: Exception) {
            mLog.error("Error opening story media stream.", e)
        }
        return@FileProxyStreamProvider null
    }

    val finalType = type
        ?: if (media.map.containsKey(MessageStoryKeys.getMessageId().instance)) PathManager.DOWNLOAD_SNAP else PathManager.DOWNLOAD_STORY
    val typeString = if (finalType == PathManager.DOWNLOAD_SNAP) "Snap" else "Story"

    context.download(
        finalType,
        mapOf("u" to media.username),
        media.extension,
        provider,
        "${media.username}'s $typeString"
    )
}

fun getMediaInfo(
    context: FeatureContext, metadata: ParamsMap, callback: (StoryMedia?) -> Unit
) {
    val map: Map<*, *> = metadata.map

    getUsername(context, metadata) { username ->
        map[StoryMetadata.getImageMediaInfo().instance]?.let {
            mLog.debug("Found image media info for story.")
            return@getUsername callback(StoryMedia(OperaMediaInfo.wrap(it), username, ".jpg", map))
        }

        map[StoryMetadata.getVideoMediaInfo().instance]?.let {
            mLog.debug("Found video media info for story.")
            return@getUsername callback(StoryMedia(OperaMediaInfo.wrap(it), username, ".mp4", map))
        }

        map[StoryMetadata.getOverlayImageMediaInfo().instance]?.let {
            mLog.debug("Found image overlay media info for story.")
            return@getUsername callback(StoryMedia(OperaMediaInfo.wrap(it), username, ".jpg", map))
        }

        mLog.error("Error getting media info for $metadata.")
        callback(null)
    }
}

private fun getUsername(
    context: FeatureContext, metadata: ParamsMap, callback: (String) -> Unit
) {
    val map: Map<*, *> = metadata.map

    val username = usernameFetchers.firstNotNullOfOrNull { it(map) }
    if (username != null) {
        mLog.debug("Found username using fetchers.")
        return callback(username)
    }

    val messageId = map[MessageStoryKeys.getMessageId().instance]
    mLog.debug("Trying to get username from message id: $messageId")
    if (messageId !is String) return callback("unknown")

    val idParts = messageId.split(':')
    if (idParts.size != 3) return callback("unknown")

    val uuid = idParts[0]
    val arroyoId = idParts[2].toLong()
    val proxy =
        MessageDummyInterface.wrap(
            MessageDummyInterface.getMappedClass().createDummyProxy(context.classLoader)
        )
    val messageCallback = DefaultFetchMessageCallback(proxy, null, 0, 0)
    context.callbacks.on(
        DefaultFetchMessageCallback::class,
        DefaultFetchMessageCallback.onFetchMessageComplete
    ) {
        val message = Message.wrap(it.args[0])
        val senderId = message.senderId.toUUIDString()
        val friends =
            context.instances.friendsRepository.selectFriendsByUserIds(listOf(senderId))
        val friendUsername =
            if (friends.isNotEmpty()) SelectFriendsByUserIds.wrap(friends[0]).username.displayString else "unknown"
        mLog.debug("Fetched message with username $friendUsername.")
        callback(friendUsername)
        true
    }
    mLog.debug("Fetching message with conversation manager.")
    context.instances.conversationManager.fetchMessage(
        uuid.toSnapUUID(),
        arroyoId,
        messageCallback
    )
}

val ParamsMap.isChat: Boolean
    get() = map.containsKey(MessageStoryKeys.getSnapInSavedState().instance)

val ParamsMap.storyId: String?
    get() {
        val reportingInfo = map[FriendStoryKeys.getStorySnapViewReportingInfo().instance]
        return if (reportingInfo is Collection<*>) StorySnapViewReportingInfo.wrap(reportingInfo.first()).storySnapKey.storyKey.storyId else null
    }

class StoryMedia(
    val info: OperaMediaInfo, val username: String, val extension: String, val map: Map<*, *>
)